### ISSUE TITLE
Classical post-processing prior to computing expval

### ISSUE THREAD

**Issue by kharazity** [PennyLaneAI/pennylane#1722]
### Feature details

For good reason, the autograd features of pennylane don't work with circuits that conclude with `qml.sample()`. However, there are many circumstances wherein one would want to do some form classical post-processing such as denoising via a neural network or symmetry post-selection prior to computing expectation values. 

Presently, there seems to be no way to perform any classical post-processing on measurement samples that allows one to use the built-in optimization libraries in pennylane. Ideally, there would be an abstraction that would allow one to evaluate a custom loss function (i.e. post-processed expval) and use the result of that loss function to optimize the underlying parameterized quantum circuit from which the samples originated. 

In particular, I am interested in the case where a neural network is trained on the sample data and is used to compute the expectation value. This seems like an important abstraction for near term devices where noise/error mitigation schemes can significantly improve algorithm performance. 

### Implementation

It seems like the implementation for expval is treated entirely by the `Device` object. As a first thought, we could write a `custom_loss` method in `Device` that takes in a lambda function and a set of arguments corresponding to arguments of the lambda. But I am not sure if this would work.

I'd be interested in learning more about the framework and implementing this as a feature if this is a feasible thing to do. 

### How important would you say this feature is?

3: Very important! Blocking work.
